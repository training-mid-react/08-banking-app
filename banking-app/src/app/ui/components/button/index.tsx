import { ClipLoader } from "react-spinners";

export const Button = (props: { label: string; isLoading: boolean }) => {
  const { label, isLoading } = props;
  return (
    <button type="submit">
      {!isLoading ? (
        label
      ) : (
        <ClipLoader
          color={"white"}
          loading={true}
          size={15}
          aria-label="Loading Spinner"
          data-testid="loader"
        />
      )}
    </button>
  );
};
