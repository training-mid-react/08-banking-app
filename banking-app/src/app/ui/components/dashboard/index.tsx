import { ROUTE_PATH } from "@core/constants";
import { useNavigate } from "react-router-dom";
import "./styles.scss";
import { IState } from "@core/interfaces";
import { FC } from "react";

export interface IDashboardProps {
  state: IState;
  handleLogout: () => void;
  refetchAccounts: () => void;
}

export const Dashboard: FC<IDashboardProps> = (props) => {
  const navigate = useNavigate();
  const { state, handleLogout, refetchAccounts } = props;

  return (
    <div className="dashboard">
      <h1 className="dashboard__title">
        Bienvenido, {state.loginData?.username}
      </h1>
      <h3 className="dashboard__balance">
        Su saldo disponible es: $
        <span className="dashboard__balance-amount">{state.balance}</span>
        <button onClick={() => refetchAccounts()}>Recargar</button>
      </h3>
      <div className="dashboard__buttons">
        <button
          className="dashboard__button dashboard__button--deposit"
          onClick={() => navigate(ROUTE_PATH.DEPOSIT)}
        >
          Depósito
        </button>
        <button
          className="dashboard__button dashboard__button--purchase"
          onClick={() => navigate(ROUTE_PATH.PURCHASE)}
        >
          Comprar
        </button>
        <button
          className="dashboard__button dashboard__button--account"
          onClick={() => navigate(ROUTE_PATH.ACOUNT)}
        >
          Gestion de cuentas
        </button>
        <button
          className="dashboard__button dashboard__button--withdraw"
          onClick={() => navigate(ROUTE_PATH.WITHDRAW)}
        >
          Retirar
        </button>
      </div>
      <div className="dashboard__logout">
        <button
          className="dashboard__button dashboard__button--logout"
          onClick={handleLogout}
        >
          Cerrar sesión
        </button>
      </div>
    </div>
  );
};
