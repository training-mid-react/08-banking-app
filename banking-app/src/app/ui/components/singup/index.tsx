import { useSignup } from "@core/hooks/signup/useSignup";
import { ROUTE_PATH } from "@core/constants";
import { Toaster } from "react-hot-toast";

export const SignupForm = () => {
  const { errors, isLoading, onSubmit, register, handleSubmit, watch } =
    useSignup();

  return (
    <div>
      <form
        id="register-form"
        className="form"
        onSubmit={handleSubmit(onSubmit)}
      >
        <h2>Regístrate</h2>
        <label>
          Nombre:
          <input
            type="text"
            placeholder="Ingrese su nombre"
            {...register("name", { required: "El nombre es obligatorio" })}
          />
          {errors.name && <span>{errors.name.message}</span>}
        </label>

        <label>
          Apellido:
          <input
            type="text"
            placeholder="Ingrese su apellido"
            {...register("lastName", {
              required: "El apellido es obligatorio",
            })}
          />
          {errors.lastName && <span>{errors.lastName.message}</span>}
        </label>

        <label>
          DNI:
          <input
            type="text"
            placeholder="Ingrese su dni"
            {...register("dni", {
              required: "El DNI es obligatorio",
            })}
          />
          {errors.dni && <span>{errors.dni.message}</span>}
        </label>

        <label>
          Nombre de usuario:
          <input
            type="text"
            placeholder="Ingrese su nombre de usuario"
            {...register("username", {
              required: "El nombre de usuario es obligatorio",
            })}
          />
          {errors.username && <span>{errors.username.message}</span>}
        </label>

        <label>
          Contraseña:
          <input
            type="password"
            placeholder="Contraseña"
            {...register("password", {
              required: "La contraseña es obligatoria",
              minLength: {
                value: 6,
                message: "La contraseña debe tener al menos 6 caracteres",
              },
            })}
          />
          {errors.password && <span>{errors.password.message}</span>}
        </label>

        <label>
          Confirmar contraseña:
          <input
            type="password"
            placeholder="Confirmar contraseña"
            {...register("confirmPassword", {
              required: "La confirmación de la contraseña es obligatoria",
              validate: (value) =>
                value === watch("password") || "Las contraseñas no coinciden",
            })}
          />
          {errors.confirmPassword && (
            <span>{errors.confirmPassword.message}</span>
          )}
        </label>

        <button type="submit" disabled={isLoading}>
          Gestion de cuentas
        </button>
        <div>
          <p>
            ¿Ya tienes cuenta?
            <a href={ROUTE_PATH.LOGIN} id="login-link">
              Inicia sesión aquí
            </a>
          </p>
        </div>
      </form>
      <Toaster />
    </div>
  );
};
