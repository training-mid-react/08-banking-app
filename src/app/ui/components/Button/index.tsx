import { FC } from 'react';
import './style.scss';

export enum ButtonType {
  SUBMIT = 'submit',
  RESET = 'reset',
  BUTTON = 'button',
}

interface ButtonProps {
  onClick: any;
  isLoading: boolean;
  className?: string;
  isDisabled: boolean;
  text: string;
  type: ButtonType;
}

export const Button: FC<ButtonProps> = ({ onClick, isLoading, className = '', isDisabled, text = '' }) => {
  return (
    <button
      className={`actionContainer__button ${className} ${isDisabled ? 'actionContainer__button--disabled' : ''} ${isLoading ? 'actionContainer__button--loading' : ''}`}
      onClick={onClick}
      disabled={isDisabled || isLoading}
      type="button"
    >
      {isLoading ? (
        <svg className="actionContainer__button-spinner" width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
          <circle cx="12" cy="12" r="10" stroke="white" strokeWidth="2" strokeLinecap="round" strokeDasharray="31.415, 31.415" strokeDashoffset="15.707" />
        </svg>
      ) : (
        text
      )}
    </button>
  );
};