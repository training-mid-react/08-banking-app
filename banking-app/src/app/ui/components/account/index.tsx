import { useCreateBankAccount } from "@core/hooks";
import { decryptAES } from "@core/utils";
import React from "react";
import { Toaster } from "react-hot-toast";
import { Button } from "../button";

export const CreateBankAccount: React.FC = () => {
  const { isLoading, accounts, errors, handleSubmit, register, onSubmit } =
    useCreateBankAccount();

  return (
    <div>
      <form onSubmit={handleSubmit(onSubmit)} className="form">
        <h2>Crear Cuenta Bancaria</h2>
        <div>
          <label htmlFor="accountType">Tipo de Cuenta</label>
          <select
            id="accountType"
            {...register("accountType", {
              required: "Selecciona un tipo de cuenta",
            })}
          >
            <option key={1} value="">
              Selecciona el tipo de cuenta
            </option>
            <option key={2} value="corriente">
              Corriente
            </option>
          </select>

          {errors.accountType && <span>{errors.accountType.message}</span>}
        </div>

        <div>
          <label htmlFor="initialBalance">Saldo Inicial ($)</label>
          <input
            type="number"
            id="initialBalance"
            placeholder="Saldo inicial"
            min={0}
            {...register("initialBalance", {
              required: "El saldo inicial es obligatorio",
              min: {
                value: 0,
                message: "El saldo inicial no puede ser negativo",
              },
            })}
          />
          {errors.initialBalance && (
            <span>{errors.initialBalance.message}</span>
          )}
        </div>
        <p>Cuentas del usuario:</p>
        {accounts.map((a) => (
          <div
            key={a.accountId}
            style={{ marginBottom: "10px", display: "flex", gap: "10px" }}
          >
            <div>Cuenta: {decryptAES(a.encryptedNumber)}</div>
            <div>Monto de la cuenta: $ {a.amount}</div>
          </div>
        ))}
        <Button label="Crear Cuenta" isLoading={isLoading} />
      </form>
      <Toaster />
    </div>
  );
};
