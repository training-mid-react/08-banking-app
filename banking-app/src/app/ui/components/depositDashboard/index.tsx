import { useDeposit } from "@core/hooks";
import { decryptAES } from "@core/utils";
import { Toaster } from "react-hot-toast";
import { Button } from "../button";

export const DepositDashboard = () => {
  const {
    isLoading,
    accounts,
    errors,
    isAccountSelected,
    handleSubmit,
    register,
    onSubmit,
  } = useDeposit();

  return (
    <div className="deposit">
      <form
        id="deposit-form"
        className="form"
        onSubmit={handleSubmit(onSubmit)}
      >
        <h2>Realizar depósito</h2>
        <label htmlFor="deposit-source">
          Selecciona el origen del depósito:
        </label>
        <select
          id="deposit-source"
          {...register("depositSource", {
            required: "El monto a depositar es obligatorio",
          })}
        >
          <option value="branch-deposit" defaultValue="branch-deposit">
            Depósito desde sucursal (Sin costo)
          </option>
          <option value="atm">
            Depósito desde cajero automático ($2 de costo)
          </option>
          <option value="account">
            Depósito desde otra cuenta ($1.5 de costo)
          </option>
        </select>
        {errors.depositSource && <span>{errors.depositSource.message}</span>}
        <label htmlFor="deposit-amount">Monto:</label>
        <input
          type="number"
          id="deposit-amount"
          placeholder="Monto a depositar"
          required
          min={0}
          {...register("amount", {
            required: "El monto a depositar es obligatorio",
            min: {
              value: 1,
              message: "El monto a depositar debe ser mayor a 0",
            },
          })}
        />
        {errors.amount && <span>{errors.amount.message}</span>}
        <div>
          <label htmlFor="accountId">Cuenta a depositar:</label>
          <select
            id="accountId"
            {...register("accountId", {
              required: "La cuenta bancaria a depositar es obligatorio",
            })}
            required
          >
            <option value="">Elige una cuenta</option>
            {accounts.map((a) => (
              <option key={a.accountId} value={a.accountId}>
                Cuenta: {decryptAES(a.encryptedNumber)} - Monto: {a.amount}
              </option>
            ))}
          </select>

          {errors.accountNumberToDeposit && (
            <span>{errors.accountNumberToDeposit.message}</span>
          )}
        </div>
        {isAccountSelected && (
          <>
            <label htmlFor="account-number-to-deposit">
              Selecciona la cuenta a depositar:
            </label>
            <select
              id="account-number-to-deposit"
              {...register("accountNumberToDeposit", {
                required: "La cuenta bancaria a depositar es obligatorio",
              })}
              required
            >
              <option value="">Elige una cuenta</option>
              {accounts.map((a) => (
                <option key={a.accountId} value={a.encryptedNumber}>
                  Cuenta: {decryptAES(a.encryptedNumber)} - Monto: {a.amount}
                </option>
              ))}
            </select>

            {errors.accountNumberToDeposit && (
              <span>{errors.accountNumberToDeposit.message}</span>
            )}
          </>
        )}
        <Button label="Depositar" isLoading={isLoading} />
        <p>Nota: El costo de la transacción se deducirá del saldo.</p>
      </form>
      <Toaster />
    </div>
  );
};
