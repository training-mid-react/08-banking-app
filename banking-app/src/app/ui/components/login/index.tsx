import { ROUTE_PATH } from "@core/constants";
import { useLogin } from "@core/hooks";
import { ILoginResponse } from "@core/interfaces";
import { Button } from "../button";

export const Login = ({
  saveLoginData,
}: {
  saveLoginData: (loginData: ILoginResponse) => void;
}) => {
  const { isLoading, errors, onSubmit, register, handleSubmit } =
    useLogin(saveLoginData);

  return (
    <div>
      <form id="login-form" className="form" onSubmit={handleSubmit(onSubmit)}>
        <h2>Iniciar sesión</h2>
        <label>
          Nombre de usuario:
          <input
            type="text"
            id="username"
            placeholder="Nombre de usuario"
            required
            {...register("username", {
              required: "El nombre de usuario es obligatorio",
            })}
          />
          {errors.username && <span>{errors.username.message}</span>}
        </label>
        <label>
          Contraseña:
          <input
            type="password"
            id="password"
            placeholder="Contraseña"
            required
            {...register("password", {
              required: "La contraseña es obligatoria",
              minLength: {
                value: 6,
                message: "La contraseña debe tener al menos 6 caracteres",
              },
            })}
          />
          {errors.password && <span>{errors.password.message}</span>}
        </label>
        <Button label="Iniciar sesión" isLoading={isLoading} />
        <p>
          ¿No tienes cuenta?
          <a href={ROUTE_PATH.SIGNUP} id="register-link">
            Regístrate aquí
          </a>
        </p>
      </form>
    </div>
  );
};
