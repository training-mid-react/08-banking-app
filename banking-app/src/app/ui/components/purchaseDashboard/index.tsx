import { Toaster } from "react-hot-toast";
import { usePurchase } from "@core/hooks";
import { decryptAES } from "@core/utils";
import { Button } from "../button";

export const PurchaseDashboard = () => {
  const { isLoading, accounts, errors, register, handleSubmit, onSubmit } =
    usePurchase();

  return (
    <div>
      <form
        id="purchase-form"
        className="form"
        onSubmit={handleSubmit(onSubmit)}
      >
        <h2>Realizar compra</h2>
        <div className="form-group">
          <label htmlFor="purchase-type">Selecciona el tipo de compra:</label>
          <select
            id="purchase-type"
            {...register("purchaseType", {
              required: "Por favor selecciona un tipo de compra",
            })}
          >
            <option value="physical">
              Compra en establecimiento físico (Sin costo)
            </option>
            <option value="online">
              Compra en línea (Costo de seguro de $5 USD)
            </option>
          </select>
          {errors.purchaseType && (
            <span className="error-text">{errors.purchaseType.message}</span>
          )}
        </div>
        <div>
          <label htmlFor="account-number-to-deposit">
            Selecciona la cuenta a depositar:
          </label>
          <select
            id="account-number-to-deposit"
            {...register("accountId", {
              required: "La cuenta bancaria a depositar es obligatorio",
            })}
            required
          >
            <option value="">Elige una cuenta</option>
            {accounts.map((a) => (
              <option key={a.accountId} value={a.accountId}>
                Cuenta: {decryptAES(a.encryptedNumber)} - Monto: {a.amount}
              </option>
            ))}
          </select>

          {errors.accountId && <span>{errors.accountId.message}</span>}
        </div>
        <div className="form-group">
          <input
            type="number"
            id="purchase-amount"
            placeholder="Monto de la compra"
            {...register("purchaseAmount", {
              required: "El monto es obligatorio",
              min: {
                value: 1,
                message: "El monto debe ser mayor que $0",
              },
            })}
          />
          {errors.purchaseAmount && (
            <span className="error-text">{errors.purchaseAmount.message}</span>
          )}
        </div>
        <Button label="Comprar" isLoading={isLoading} />
      </form>

      <Toaster />
    </div>
  );
};
