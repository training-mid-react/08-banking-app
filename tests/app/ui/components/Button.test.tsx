import React from 'react';

import { fireEvent, render, screen } from '@testing-library/react';
import { Button } from '../../../../src/app/ui/components/Button';

describe('Pruebas en el componente Button', () => {
  test('Match snapshot', () => {
    const { asFragment } = render(<Button text="Hola mundo" handleClick={() => { }} type='button' />);
    expect(asFragment()).toMatchSnapshot();
  });
  
  test('Renderiza el texto correctamente', () => {
    render(<Button text="Hola mundo" handleClick={() => { }} type='button' />);    
    expect(screen.getByText('Hola mundo')).toBeTruthy();
  });

  test('Llama la funcion handleClick correctamente', () => {
    const handleClick = vi.fn();
    render(<Button text="Hola mundo" handleClick={handleClick} type='button' />);
    const button = screen.getByTestId('button');
    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});