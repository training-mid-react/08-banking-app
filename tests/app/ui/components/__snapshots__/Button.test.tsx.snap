// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pruebas en el componente Button > Match snapshot 1`] = `
<DocumentFragment>
  <button
    data-testid="button"
    type="button"
  >
    Hola mundo
  </button>
</DocumentFragment>
`;
